### ISSUE TITLE
[Feature Request] Globally Suppress ANSI Sequences

### ISSUE THREAD

**Issue by KyleKing** [beartype/beartype#201]
For applications, it would be nice if the color setting could be globally configurable by an environment variable. Related to #162 and #178

Maybe `export BEARTYPE_NO_COLOR=1` like [better-exceptions docs](https://github.com/qix-/better-exceptions#usage) and [better_exceptions/color.py#L90](https://github.com/Qix-/better-exceptions/blob/3b6c15699b6343459c5852c3f748630afb8f43ef/better_exceptions/color.py#L90)

```py
os.getenv("BEARTYPE_NO_COLOR") or
conf.is_color is False or
(conf.is_color is None and not is_stdout_terminal())
```

From:
https://github.com/beartype/beartype/blob/308925d9e1b9e1e69886e0ed6904648612b01e7d/beartype/_decor/_error/_util/errorutilcolor.py#L179-L193

The caveat is that this would be the solitary environment configuration and there might be a better way for `beartype` to support global configuration

I have a particular issue with structlog "rich" exceptions printing the ANSI codes without formatting, but more generally there have been issues with the ANSI codes appearing in stdout

**Comment by leycec** [comment]
**Excellent idea!** Sadly, this is unlikely to make it in time for our upcoming **@beartype 0.12.0** release. Happily, you *should* be able to roll your own `@beartype_monochrome` decorator based on an app-specific `BEARTYPE_NO_COLOR` environment variable in the meantime:

```python
from beartype import BeartypeConf, beartype
from os import getenv

beartype_monochrome = (
    beartype(conf=BeartypeConf(is_color=False))
    if os.getenv("BEARTYPE_NO_COLOR") else
    beartype
)
'''

Thanks so much for the superb request. If I fail to do this, please prod me with rusty poker until I eventually relent. :grin:

**Comment by KyleKing** [comment]
I think this issue can be closed now that globally configuring `beartype` is the future!

![Akira gif](http://images6.fanpop.com/image/photos/37900000/Gifs-from-Akira-anime-37997391-300-161.gif)

```py
...

beartype_this_package(conf=BeartypeConf(
    is_color=getenv('BEARTYPE_NO_COLOR') is not None),
)
```

https://github.com/KyleKing/calcipy_template/blob/23fa65c0400cac50c1cc1b3770526af669ce0159/package_template/%7B%7B%20package_name_py%20%7D%7D/__init__.py.jinja#L39-L48

**Comment by leycec** [comment]
**Gah!** Jaw-dropping *Akira* wheely scene requires I finally do something about this. Ordinarily, I'd be inclined to fully agree that I should do nothing. That's the path of lazy resistance. But...

`calcipy_template` (and therefore possibly other third-party packages) are now claiming ownership over environment variables prefixed by `BEARTYPE_` like `BEARTYPE_NO_COLOR`, which... isn't great. It isn't great for @beartype, it isn't great for third-party packages, and it isn't great for forward compatibility between @beartype and third-party packages.

It's possible – nay, even likely – that some future release of @beartype will want to standardize environment variables prefixed by `BEARTYPE_`. That release is now!

## Let @leycec Tell Everybody What Now

`BEARTYPE_NO_COLOR` is compelling – but I think @beartype can do sorta better. Instead of a boolean `BEARTYPE_NO_COLOR`, our upcoming `beartype 0.15.1` release will support a new tri-state `BEARTYPE_IS_COLOR` environment variable whose value *must* be one of three string literals:

* `"True"`, corresponding to `BeartypeConf(is_color=True)`.
* `"False"`, corresponding to `BeartypeConf(is_color=False)`.
* `"None"`, corresponding to `BeartypeConf(is_color=None)`.

If defined, this environment variable **will take precedence over any explicitly passed `BeartypeConf(is_color=...)` parameter.** Specifically, if both the `BEARTYPE_IS_COLOR` environment variable is defined *and* the `BeartypeConf(is_color=...)` parameter is passed, then @beartype will:

* Emit a non-fatal warning that the passed `BeartypeConf(is_color=...)` parameter is being ignored.
* Override the value of the passed `BeartypeConf(is_color=...)` parameter with that of the `BEARTYPE_IS_COLOR` environment variable.

## I Come to You and Ask for a Favor, @KyleKing

In return, would you mind renaming (or possibly entirely dropping) your existing `BEARTYPE_NO_COLOR` parameter from `calcipy_template`? I'd *really* like to avoid a scenario where third-party authors begin claiming ownership over environment variables prefixed by `BEARTYPE_`.

Clearly, it's impossible to prevent that. Clearly, a bear bro has gotta try!

*Do it for Akira.*

![for akira, bro](https://media.tenor.com/srx88YYGl9sAAAAC/akira-motorcycle.gif)

**Comment by KyleKing** [comment]
Bearly avoiding infinite variations of environment variables seems like a good plan to me!

I went ahead and updated the template repository, which for all my effort is only just me 😅

**Comment by leycec** [comment]
Thanks for being so accommodating, and... sorry about all this feature churn. I tell myself in the voluminous dark of the Canadian night that API instability is why I *still* haven't officially documented `beartype.claw`. In reality, it's because I'm bone weary. :face_with_spiral_eyes: 

Oh, and I'm certain you'll see an uptick in the `calcipy` userbase! Awesome software attracts awesome users. Have you considered pushing `calcipy` on geek-oriented social democracy platforms like [/r/Python](https://www.reddit.com/r/Python/) and [Hacker News](https://news.ycombinator.com/)? I *very* occasionally submit @beartype links to both – and always see a dramatic increase in GitHub stars in the immediate aftermath. It's kinda fun to show the world all the nifty stuff we do, too. Man can only code in his dank den of isolation for so long before the jitters set in.

> "Code it and they will come." — Kevin Costner's brother, probably.

**Comment by leycec** [comment]
**Behold!** @leycec did something. By which I mean... **resolved by 547f37e.**

[Read all about it at our official ReadTheDocs (RTD)-hosted documentation](https://beartype.readthedocs.io/en/latest/api_decor/#beartype-is-color). Thanks for *bearing* with us, @KyleKing. It may take us a year to resolve a single feature request, but we did the right thing in the end. The world is your oyster, `${BEARTYPE_IS_COLOR}`!
